### DISCUSSION TITLE
Exception during no records in the database with "yeild db" in the get_db()

### DISCUSSION THREAD

**Discussion by vinayvinu500** [fastapi/fastapi#11297] [Questions] [answered]
### First Check

- [x] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [ ] I searched the FastAPI documentation, with the integrated search.
- [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [ ] I already read and followed all the tutorial in the docs and didn't find an answer.
- [ ] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# main.py file 
from fastapi import FastAPI, Depends, HTTPException, status
from sqlalchemy.orm import Session
from databases import engine, SessionLocal, get_db
from models import Base, Users
from schemas import POST_USERS, GET_USERS
from typing import Annotated

app = FastAPI()
Base.metadata.create_all(bind=engine)


# Start of the database.py file
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

SQLALCHEMY_DATABASE_URI = "sqlite:///sqlite.db"
engine = create_engine(SQLALCHEMY_DATABASE_URI, connect_args={'check_same_thread': False})
SessionLocal = sessionmaker(bind=engine, autoflush=False, autocommit=False)

async def get_db():
    db = SessionLocal()
    try:
        yield db
    except Exception as e:
        print(f'There was a error in the database: {e}')
    finally:
        db.close()

# end of the database.py file

# start of the model.py file
from sqlalchemy import Column, Integer, String, Boolean, ForeignKey
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()

class Users(Base):
    __tablename__ = 'users'
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    username = Column(String, unique=True, nullable=False)
    password = Column(String, nullable=False)
    email = Column(String, unique=True, nullable=False)

# end of the models.py file

# start of the schemas.py file
from pydantic import BaseModel

# POST 
class POST_USERS(BaseModel):
    username: str
    password: str
    email: str

# GET
class GET_USERS(BaseModel):
    username: str
    email: str

    class Config():
        # orm_mode = True # UserWarning: Valid config keys have changed in V2: * 'orm_mode' has been renamed to 'from_attributes'
        from_attributes = True

# end of the schemas.py file

@app.get('/users', response_model=list[GET_USERS])
def get_users(db: Annotated[Session, Depends(get_db)]):
    users_db = db.query(Users).all()
    if users_db in ([], None):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail='There are no records in the database!') 
        """
        INFO:     127.0.0.1:50292 - "GET /docs HTTP/1.1" 200 OK
        INFO:     127.0.0.1:50292 - "GET /openapi.json HTTP/1.1" 200 OK
        There was a error in the database: 404: There are no records in the database!
        INFO:     127.0.0.1:50115 - "GET /users HTTP/1.1" 500 Internal Server Error
        ERROR:    Exception in ASGI application
        raise FastAPIError(fastapi.exceptions.FastAPIError: No response object was returned. 
        There's a high chance that the application code is raising an exception and a dependency with yield has a block with a bare except, 
        or a block with except Exception, and is not raising the exception again.
        Read more about it in the docs: https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#dependencies-with-yield-and-except
        """

    return users_db  

# requirements.txt 
"""
fastapi
sqlalchemy
uvicorn
"""
```

### Description

Thank you for the wonderful documentation and got missed with one point where initially the database will be not having any records and eventually we get the result as a empty list when we fetch the records using sqlalchemy orm and in the if statement we will raise HTTPException() or possibly return the message in the dict format {"message": "There are no records in the database!"}

but the real problem is when you will be returning the message or raise the exception the message in the detail parameter 
it will be catching the exception in the get_db() function instead of the returning the message
where as there will be "500 Internal Server Error" in the terminal and to the end user

Error was 
"""
        INFO:     127.0.0.1:50292 - "GET /docs HTTP/1.1" 200 OK
        INFO:     127.0.0.1:50292 - "GET /openapi.json HTTP/1.1" 200 OK
        There was a error in the database: 404: There are no records in the database!
        INFO:     127.0.0.1:50115 - "GET /users HTTP/1.1" 500 Internal Server Error
        ERROR:    Exception in ASGI application
        raise FastAPIError(fastapi.exceptions.FastAPIError: No response object was returned. 
        There's a high chance that the application code is raising an exception and a dependency with yield has a block with a bare except, 
        or a block with except Exception, and is not raising the exception again.
        Read more about it in the docs: https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#dependencies-with-yield-and-except
"""

for more info, please do check-out my repo: https://github.com/vinayvinu500/Training 
![Users - Get Request - Internal Server Error](https://github.com/tiangolo/fastapi/assets/49785350/16795d34-1aa5-45aa-bcf9-5f731a6cc8cb)

### Operating System

Windows

### Operating System Details

Processor	11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz   2.42 GHz
Installed RAM	16.0 GB (15.7 GB usable)
System type	64-bit operating system, x64-based processor

### FastAPI Version

0.110.0

### Pydantic Version

2.6.3

### Python Version

Python 3.11.4

### Additional Context

Instead of Exception I need to return the message as it is in the detail parameter of the `HTTPException`

**Comment by acidjunk** [comment]
Why is no users an error status code in the 400 range?
That's normally for a client side error.

**Comment by JavierSanchezCastro** [comment, marked as answer]
In the latest version (0.110.0) a breaking change has been included which includes exception handling via yield https://fastapi.tiangolo.com/release-notes/#breaking-changes.

To return the message you want, you need to add a raise in the exception block, like:

```python3
async def get_db():
    db = SessionLocal()
    try:
        yield db
    except Exception as e:
        print(f'There was a error in the database: {e}')
        raise
    finally:
        db.close()
```

**Reply by vinayvinu500** [reply]
```
async def get_db():
    db = LocalSession()
    try: 
        yield db
    except Exception as e:
        print("There is a error in database instance: %s" % e)
        raise HTTPException(status_code=404, detail=f"{e}")
    finally:
        db.close()
```

instead of the above only raise, I have raised again with the same error code 
I don't know, Is this a standard practice or not because as catching some error instance in the Except block and again raise is bad practice I suppose...

![image](https://github.com/tiangolo/fastapi/assets/49785350/57ece3f5-a9c6-4b9a-90d2-1a697c49b9ba)

**Reply by JavierSanchezCastro** [reply]
You only need to add `raise` after your print (not with HTTPException). Following the docs, it's to avoid memory leaks.

You can get more info from https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#dependencies-with-yield-and-except

**Reply by vinayvinu500** [reply]
Thanks @JavierSanchezCastro this serves the purpose...

**Comment by vinayvinu500** [comment]
I Found out good Solution in which we are only sharing the message instead of error or raise the error

```
from typing import Union, List
class Message(BaseModel):
    message: str

@app.get('/users', response_model=Union[List[GET_USERS], Message])
def get_users(db: Annotated[Session, Depends(get_db)]):
    users_db = db.query(Users).all()
    if users_db in ([], None):
        return {"message": "There are no records in the database!"}

    return users_db  
```


Please do let me know is this a standard practice to return the message through Pydantic way!
or the just raise the error as suggested by [JavierSanchezCastro](https://github.com/JavierSanchezCastro)
![image](https://github.com/tiangolo/fastapi/assets/49785350/9d36900c-6bdc-4417-8ac1-fa6c2f84a6ea)
